lib/active_storage/blob_io.rb

```diff
@@ -0,0 +1,54 @@
+module FormatParser
+  module ActiveStorage
+    class BlobIO
+      def initialize(blob)
+        @blob = blob
+        @service = blob.service
+        @pos = 0
+        @remote_size = false
+      end
+
+      # Emulates IO#seek
+      def seek(offset)
+        @pos = offset
+        0 # always return 0
+      end
+
+      # Emulates IO#pos
+      def pos
+        @pos
+      end
+
+      # Emulates IO#size.
+      def size
+        raise 'Remote size not yet obtained, need to perform at least one read() to retrieve it' unless @remote_size
+        @remote_size
+      end
+
+      def read(n_bytes)
+        range_start = @pos
+        range_end =  @pos + n_bytes - 1
+        http_range = (range_start..range_end)
+
+        # ActiveStorageAdapter#download_chunk method return '' when using range (start > end)
+        # Raise invalid range exception if this is the case to recognise the real empty response
+        raise ArgumentError, "Invalid range from #{range_start} to #{range_end}" if range_start > range_end
+
+        begin
+          body = @service.download_chunk(@blob.key, http_range)
+        rescue Errno::ENOENT
```

Comment:
Are you sure these are the exceptions that the adapters will raise? Do they convert invalid range HTTP code (416 afaik) to EINVAL and ENOENT? Also, the S3 endpoints will give you a 403 (access denied) instead of ENOENT if you try to request from a nonexistent resource unless you happen to have a "List objects" permission on the bucket. How do the service adapters in ActiveStorage deal with these cases?

Reply:
Yes, these exceptions are the ones being raised for DiskStorage, will try to check for [other adapters](https://github.com/rails/rails/tree/master/activestorage/lib/active_storage/service) as well.